tiny_dnn/layers/layer.h

```diff
@@ -454,6 +458,52 @@ class layer : public node {
     }
   }
 
+#ifdef CNN_USE_HDF
```

Comment:
provide this as an example/tutorial. Map data to `std::vector<Tensor<>>` and then overload operators in network, later and parameter to make a tiny API in order to load the data

Reply:
This sounds good. I'll do it here.

Reply:
We should take this project as a reference and try to load that networks in a similar way
https://github.com/szagoruyko/functional-zoo

See that the parameters are fetched to the layers independently to the serializer. It uses numpy array as an intermediate representation to map to `torch.Tensor` , in our case will be directly `std::vector<float>` to `dnn::Tensor`

Reply:
So shall we have a `hdf_serializer` class in a file under `tiny_dnn/util` which does this tying of data?

Reply:
And I think if we wish to do provide framework specific support for special cases, we can define other serializers inheriting this one. And the API can be much similar like `optimizers` and `parameter_init`.

Just as `parameter.initialize` takes an argument as `parameter_init::function`, we can do `parameter.load(hdf_serializers::serializer &s, args... )` and it depends on use which serializer to use (in his own program).

Reply:
We shall have support but internally we will be independent of the file structure and version hassles. What say ?
/cc: @bhack @nyanp @Randl @prlz77 you are welcome to join the discussion :smile:

Reply:
That approach sounds better but I guess it might take some time to refactor the serializer. I'd go first to the simplest case and load data assuming that parsing is done by the user.

Reply:
@karandesai-96 
I think there are 4 different abstraction levels for importer API in this discussion:

1. Tensor API (Just loading tensor from file, as @edgarriba said)
1. Layer API (#868)
1. Network API (#869)
1. Model API (#870)

Supporting 1 and 4 are good enough for me (providing both flexibility and convenience). But if we provides 2 and 3 as public API, we will need to make a lot of effort to support various user-defined models. In addition to that, we need to write additional code if we add a new layer into tiny-dnn.

Constructing Model API over primitive tensor API seems good for maintainability point of view.

Reply:
@nyanp @edgarriba From the last GoogleBrain Ama [seems](https://www.reddit.com/r/MachineLearning/comments/6z51xb/we_are_the_google_brain_team_wed_love_to_answer/#dmtlgdt) that TF team is open on [onnx](https://github.com/onnx/onnx). What do you think?

Reply:
@bhack /cc @edgarriba   @karandesai-96
ONNX is very promising format if TF is open on it. This is one of the reason why I don't want to provide layer/network API with Keras's native format. After supporting popular model with minimum effort, I want to support ONNX (it seems good path for us to survive).

Reply:
agree !